src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -1580,237 +1034,195 @@ void hashForModule(String name, byte[] hash) {
-                    // if non-empty and more than one set reference this builder,
-                    // emit to a unique local
-                    index = refCount <= 1 ? defaultVarIndex
-                                          : nextLocalVar.getAsInt();
-                    if (index < MAX_LOCAL_VARS) {
-                        localVarIndex = index;
+            class CacheBuilder {
+                private static final String VALUES_ARRAY = "dedupSetValues";
+                final ArrayList<Snippet> cachedValues = new ArrayList<>();
+
+                // Load the set to the operand stack.
+                // When referenced more than once, the value is pre-built with static initialzer
+                // and is load from the cache array with
+                //   dedupSetValues[index]
+                // Otherwise, LoadableSet will load the set onto the operand stack.
+                private Snippet loadFromCache(int index) {
+                    assert index >= 0;
+                    return cob ->
+                        cob.getstatic(owner, VALUES_ARRAY, CD_Set.arrayType())
+                           .loadConstant(index)
+                           .aaload();
+                }
+
+                Snippet transform(SetReference<?> setRef, Snippet loadSnippet) {
+                    if (setRef.refCount() > 1) {
+                        cachedValues.add(loadSnippet);
+                        return loadFromCache(cachedValues.size() - 1);
                     } else {
-                        // overflow: disable optimization by using localVarIndex = 0
-                        index = defaultVarIndex;
+                        return loadSnippet;
                     }
-
-                    generateSetOf(cob, index);
                 }
-                return index;
-            }
 
-            private void generateSetOf(CodeBuilder cob, int index) {
-                if (elements.size() <= 10) {
-                    // call Set.of(e1, e2, ...)
-                    for (T t : sorted(elements)) {
-                        visitElement(t, cob);
+                /*
+                * Adding provider methods to the class. For those set used more than once, built
+                * once and keep the reference for later access.
+                * Return a snippet to setup the cache <clinit>.
+                *
+                * The returned snippet would set up the set referenced more than once,
+                *
+                * static final Set[] dedupSetValues;
+                *
+                * static {
+                *     dedupSetValues = new Set[countOfStoredValues];
+                *     dedupSetValues[0] = Set.of(elements); // elements no more than SET_SIZE_THRESHOLD
+                *     dedupSetValues[1] = dedup<setWithIndex>Provider(); // set elements more than SET_SIZE_THRESHOLD
+                *     ...
+                *     dedupSetValues[countOfStoredValues - 1] = ...
+                * }
+                */
```

Comment:
This javadoc also needs update.   The helper method name no longer uses "Provider" suffix.